TombEngine/Objects/TR3/Trap/ElectricCleaner.cpp

```diff
@@ -0,0 +1,476 @@
+#include "framework.h"
+#include "Objects/TR3/Trap/ElectricCleaner.h"
+
+#include "Game/collision/collide_item.h"
+#include "Game/control/box.h"
+#include "Game/effects/debris.h"
+#include "Game/effects/effects.h"
+#include "Game/effects/item_fx.h"
+#include "Game/effects/spark.h"
+#include "Game/effects/tomb4fx.h"
+#include "Game/Lara/lara_helpers.h"
+#include "Math/Math.h"
+#include "Specific/level.h"
+#include "Specific/setup.h"
+
+using namespace TEN::Effects::Items;
+using namespace TEN::Effects::Spark;
+
+// ItemFlags[0]:	   Rotation speed and heading angle.
+// ItemFlags[1]:	   Has reached a new sector and can unblock the one behind it. Unused but reserved just in case.
+// ItemFlags[2]:	   Movement velocity.
+// ItemFlags[3, 4, 5]: Counters for dynamic lights and sparks.
+
+// OCB:
+// 0:			  Stop after killing the player.
+// Anything else: Don't stop after killing the player.
+
+namespace TEN::Entities::Traps
+{
+	constexpr auto ELECTRIC_CLEANER_VELOCITY  = BLOCK(1 / 16.0f);
+	constexpr auto ELECTRIC_CLEANER_TURN_RATE = 1024;
+
+	const auto ElectricCleanerHarmJoints = std::vector<unsigned int>{ 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13 };
+
+	void InitialiseElectricCleaner(short itemNumber)
+	{
+		auto& item = g_Level.Items[itemNumber];
+
+		// Align to the middle of the block.
+		item.Pose.Position.x = (item.Pose.Position.x & ~WALL_MASK) | (int)BLOCK(0.5f);
+		item.Pose.Position.z = (item.Pose.Position.z & ~WALL_MASK) | (int)BLOCK(0.5f);
+
+		// Init flags.
+		item.ItemFlags[0] = ELECTRIC_CLEANER_TURN_RATE;
+		item.ItemFlags[1] = 0;
+		item.ItemFlags[2] = ELECTRIC_CLEANER_VELOCITY;
+		item.Collidable = 1;
+	}
+
+	void ElectricCleanerControl(short itemNumber)
+	{
+		auto& item = g_Level.Items[itemNumber];
+
+		if (!TriggerActive(&item))
+		{
+			if (item.ItemFlags[2])
+			{
+				item.ItemFlags[2] = 0;
+				SoundEffect(SFX_TR3_CLEANER_FUSEBOX, &item.Pose);
+			}
+
+			return;
+		}
+
+		if (!item.ItemFlags[2])
+			return;
+
+		// Not facing a quadrant; keep turning.
+		if (item.Pose.Orientation.y & 0x3FFF)
+		{
+			item.Pose.Orientation.y += item.ItemFlags[0];
+		}
+		else
+		{
+			// Do we need a new target?
+			if (NeedNewTarget(item))
+			{
+				long x, y, z;
+				FloorInfo* floor;
+				ROOM_INFO* r;
+				short roomNumber;
+
+				if (item.ItemFlags[1])
+				{
+					x = item.Pose.Position.x + (BLOCK(1) * phd_sin(item.Pose.Orientation.y + ANGLE(180.0f)));
+					y = item.Pose.Position.y;
+					z = item.Pose.Position.z + (BLOCK(1) * phd_cos(item.Pose.Orientation.y + ANGLE(180.0f)));
+					roomNumber = item.RoomNumber;
+					floor = GetFloor(x, y, z, &roomNumber);
+					r = &g_Level.Rooms[roomNumber];
+					r->floor[((z - r->z) / BLOCK(1)) + r->xSize * ((x - r->x) / BLOCK(1))].Stopper = 0;
+					item.ItemFlags[1] = 0;
+				}
+
+				// Check where cleaner is heading and determine where to go next.
+				bool left, ahead;
+				y = item.Pose.Position.y;
+
+				switch (item.Pose.Orientation.y)
+				{
+				// Facing Z+
+				case ANGLE(0.0f):
```

Comment:
I'm seeing this switch is checking each cardinal point, but except few changes, most of the code inside are clones. 
Maybe this could be cleaner using a sub-function to reduce the number of cloned codes.

Still, not a prioritary issue. So if decided to leave as it's. That's fine for me too.

Reply:
Yes, I wanted to decopypaste it, but I did not have enough time. I will be grateful if someone else takes this task.

Reply:
Done, I reworked the code in this other branch: https://github.com/MontyTRC89/TombEngine/pull/1005

Part of the new editions involves a new code flow that avoid duplicated rutines.